music/src/music_vae/model.ts

```diff
@@ -71,11 +71,65 @@ abstract class Encoder {
 }
 
 /**
- * A single-layer bidirectional LSTM encoder.
+ * A single-layer bidirectional LSTM module.
  */
-class BidirectionalLstmEncoder extends Encoder {
+class BidirectionalLstm {
   private lstmFwVars: LayerVars;
   private lstmBwVars: LayerVars;
+
+  /**
+   * `BidirectionalLstm` contructor.
+   *
+   * @param lstmFwVars The forward LSTM `LayerVars`.
+   * @param lstmBwVars The backward LSTM `LayerVars`.
+   */
+  constructor(lstmFwVars: LayerVars, lstmBwVars: LayerVars) {
+    this.lstmFwVars = lstmFwVars;
+    this.lstmBwVars = lstmBwVars;
+  }
+
+  /**
+   * Processes a batch of sequences.
+   * @param sequence The batch of sequences to be processed.
+   * @returns A batch of forward and backward output LSTM states.
+   */
+  process(sequence: tf.Tensor3D): [tf.Tensor2D[], tf.Tensor2D[]] {
+    return tf.tidy(() => {
+      const fwStates = this.singleDirection(sequence, true);
+      const bwStates = this.singleDirection(sequence, false);
+      return [fwStates, bwStates];
+    });
+  }
+
+  private singleDirection(inputs: tf.Tensor3D, fw: boolean) {
+    const batchSize = inputs.shape[0];
+    const length = inputs.shape[1];
+
+    const lstmVars = fw ? this.lstmFwVars : this.lstmBwVars;
+    let state: [tf.Tensor2D, tf.Tensor2D] = [
+      tf.zeros([batchSize, lstmVars.bias.shape[0] / 4]),
+      tf.zeros([batchSize, lstmVars.bias.shape[0] / 4])
+    ];
+    const forgetBias = tf.scalar(1.0);
+    const lstm = (data: tf.Tensor2D, state: [tf.Tensor2D, tf.Tensor2D]) =>
+        tf.basicLSTMCell(
+            forgetBias, lstmVars.kernel, lstmVars.bias, data, state[0],
+            state[1]);
+    const splitInputs = tf.split(inputs.toFloat(), length, 1);
+    const outputStates: tf.Tensor2D[] = [];
+    for (const data of (fw ? splitInputs : splitInputs.reverse())) {
+      state = lstm(data.squeeze([1]) as tf.Tensor2D, state);
+      outputStates.push(state[1]);
```

Comment:
maybe we should call this outputStatesH since we're not keeping the C?

Reply:
Actually we can also just document this better ... like above we can say "A batch of forward and backward output (h) LSTM states."

Reply:
My LSTM recollection is rusty but "h" is the same as "output", right?  And "c" is the internal state?  I'll add some comments.  BTW I didn't modify this code at all, just refactored so I could use the same bidirectional LSTM code to preprocess control sequences.

Reply:
yeah h is output -- but putting "(h)" will just make it more explicit  -- i think just adding that is enough

Reply:
Added that and a tiny bit more.